src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -241,6 +280,240 @@ void DiscoveryDataBase::AckedFunctor::operator () (
+        {
+            if(rit->guidPrefix == participant_guid.guidPrefix)
+            {
+                tit->second.erase(rit);
+                --rit;
+            }
+        }
+    }
+
+    for (auto tit = writers_by_topic_.begin(); tit != writers_by_topic_.end(); ++tit)
+    {
+        for (auto wit = tit->second.begin(); wit != tit->second.end(); ++wit)
+        {
+            if(wit->guidPrefix == participant_guid.guidPrefix)
+            {
+                tit->second.erase(wit);
+                --wit;
+            }
+        }
+    }
+
+    // Remove participant from others participants_[]::relevant_participants_builtin_ack_status
+    for (auto pit = participants_.begin(); pit != participants_.end(); ++pit)
+    {
+        pit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    // Remove participant from others writers_[]::relevant_participants_builtin_ack_status
+    for (auto wit = writers_.begin(); wit != writers_.end(); ++wit)
+    {
+        wit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    // Remove participant from others readers_[]::relevant_participants_builtin_ack_status
+    for (auto pit = readers_.begin(); pit != readers_.end(); ++pit)
+    {
+        pit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    if(std::find(disposals_.begin(), disposals_.end(), ch) == disposals_.end())
+    {
+        disposals_.push_back(ch);
+    }
+}
+
+void DiscoveryDataBase::create_participant_from_change(
+        eprosima::fastrtps::rtps::CacheChange_t* ch)
+{
+    DiscoveryParticipantInfo part(ch);
+    participants_.insert(std::make_pair(guid_from_change(ch).guidPrefix, part));
+}
+
+void DiscoveryDataBase::create_writers_from_change(
+        eprosima::fastrtps::rtps::CacheChange_t* ch,
+        const eprosima::fastrtps::string_255& topic_name)
+{
+    const eprosima::fastrtps::rtps::GUID_t& writer_guid = guid_from_change(ch);
+
+    DiscoveryEndpointInfo tmp_writer(ch, topic_name);
+    std::pair<std::map<eprosima::fastrtps::rtps::GUID_t,DiscoveryEndpointInfo>::iterator,bool> ret;
+    ret = writers_.insert(std::make_pair(writer_guid, tmp_writer));
+
+    if (ret.second) {
+        auto readers_it = readers_by_topic_.find(topic_name);
+        if (readers_it != readers_by_topic_.end())
+        {
+            for (auto reader_it: readers_it->second)
+            {
+                writers_.find(writer_guid)->second.add_or_update_ack_participant(reader_it.guidPrefix);
+
+                auto rit = readers_.find(reader_it);
+                if (rit != readers_.end())
+                {
+                    rit->second.add_or_update_ack_participant(writer_guid.guidPrefix);
+                }
+
+                auto pit = participants_.find(reader_it.guidPrefix);
+                if (pit != participants_.end())
+                {
+                    pit->second.add_or_update_ack_participant(writer_guid.guidPrefix);
+                }
```

Comment:
I think this part isn't necessary. The status in `participants_` is for the acknowledgement of the DATA(p). Doing this will set the ack status of the PDP of the writer in the PDP of the reader to false even if they knew each other beforehand. You could check whether is matched before adding:
```suggestion
                auto pit = participants_.find(reader_it.guidPrefix);
                if (pit != participants_.end())
                {
                    if (!pit->second.is_matched(writer_guid.guidPrefix))
                    {
                        pit->second.add_or_update_ack_participant(writer_guid.guidPrefix);
                    }
                }
```